core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala

```diff
@@ -254,53 +258,65 @@ private[spark] class CoarseMesosSchedulerBackend(
+              taskIdToSlaveId.put(taskId, slaveId)
+              slaveIdsWithExecutors += slaveId
+              coresByTaskId(taskId) = cpusToUse
+              // Gather cpu resources from the available resources and use them in the task.
+              val (remainingResources, cpuResourcesToUse) =
+                partitionResources(offer.getResourcesList, "cpus", cpusToUse)
+              val (_, memResourcesToUse) =
+                partitionResources(remainingResources.asJava, "mem", calculateTotalMemory(sc))
+              val taskBuilder = MesosTaskInfo.newBuilder()
+                .setTaskId(TaskID.newBuilder().setValue(taskId.toString).build())
+                .setSlaveId(offer.getSlaveId)
+                .setCommand(createCommand(offer, cpusToUse + extraCoresPerSlave, taskId))
+                .setName("Task " + taskId)
+                .addAllResources(cpuResourcesToUse.asJava)
+                .addAllResources(memResourcesToUse.asJava)
+
+              sc.conf.getOption("spark.mesos.executor.docker.image").foreach { image =>
+                MesosSchedulerBackendUtil.setupContainerBuilderDockerInfo(image, sc.conf,
+                  taskBuilder.getContainerBuilder())
+              }
+
+              // Accept the offer and launch the task
+              logDebug(s"Accepting offer: $id with attributes: $offerAttributes" +
+                s" mem: $mem cpu: $cpus")
+              slaveIdToHost(offer.getSlaveId.getValue) = offer.getHostname
+              d.launchTasks(
+                Collections.singleton(offer.getId),
+                Collections.singleton(taskBuilder.build()), filters)
+            } else {
+              // Decline the offer
+              logDebug(s"Declining offer: $id with attributes: $offerAttributes" +
+                s" mem: $mem cpu: $cpus")
+              d.declineOffer(offer.getId)
             }
-
-            // Accept the offer and launch the task
-            logDebug(s"Accepting offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
-            slaveIdToHost(offer.getSlaveId.getValue) = offer.getHostname
-            d.launchTasks(
-              Collections.singleton(offer.getId),
-              Collections.singleton(taskBuilder.build()), filters)
           } else {
-            // Decline the offer
-            logDebug(s"Declining offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
-            d.declineOffer(offer.getId)
+            // We reached the maximum number of cores for this framework. We don't need to see
+            // new offers. Decline the offer for a long period of time.
+            logDebug(s"Declining offer (reached max cores): $id with attributes:" +
+              s" $offerAttributes mem: $mem cpu: $cpus" +
+              s" for $rejectOfferDurationForReachedMaxCores seconds")
+            d.declineOffer(offer.getId, Filters.newBuilder()
+              .setRefuseSeconds(rejectOfferDurationForReachedMaxCores).build())
```

Comment:
It also depends on `id`, `offerAttributes`, `mem` and `cpus` for logging. They're all derived from `offer` and some are easier than others to get but we shouldn't compute them twice just for logging.

Okay I'll change it to handle only the case where offers are rejected for a longer period.